### PR TITLE
Notifications form

### PR DESCRIPTION
**Pull request by keithjgrant**
##### SUMMARY
Adds notification template form and add/edit pages

addresses #7876 

Note: this probably needs a touch more polish but I wanted to get some eyes on it while I'm out on PTO next week

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
 - UI

### CODE REVIEW

**Comment by AlexSCorey** [comment]
@keithjgrant 
Just poking at this a bit:
1) I was able to create an email notification, but then I edited it, and removed the recipients list (which says it's required) and I was able to still save.  However, it did not save the removed, recipients list
2) On the list, it seems like in most places where we have an inline label we are not using a colon to separate the label from the value.  We do use a colon in Survey List, but that's the only place that I found it in a brief search.  @trahman73 can you comment on whether to use the colon?
3) Disable the save button.  This is a new pattern that we need the implement here and in all other forms. (https://github.com/ansible/awx/issues/7773)

**Comment by trahman73** [comment]
@AlexSCorey That's correct, we do not use a colon to separate a label from it's value.

File: awx/ui_next/src/screens/NotificationTemplate/shared/CustomMessagesSubForm.jsx [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+        onChange={() => useCustomHelpers.setValue(!useCustomField.value)}
+      />
+      {useCustomField.value && (
+        <SubFormLayout>
+          <Text
+            className="pf-c-content"
+            css="margin-bottom: var(--pf-c-content--MarginBottom)"
+          >
+            <small>
+              Use custom messages to change the content of notifications sent
```
Comment (mabashian) [inline awx/ui_next/src/screens/NotificationTemplate/shared/CustomMessagesSubForm.jsx]: Hrmmm we'll have to figure out how to translate all of this

**Comment by mabashian** [comment]
@keithjgrant I went ahead and rebased this to make it a bit easier to review.  I noticed that there were some failing tests in the cred form test suite so I pushed a fix for that.  It looks like the only thing left that's failing in the awx-ui-next is a lint failure:

```
2020-08-26 14:15:03.399065 | static | /awx_devel/awx/ui_next/src/components/FormField/ArrayTextField.jsx
2020-08-26 14:15:03.399143 | static |   43:19  error  'value' is already declared in the upper scope  no-shadow
```

**Comment by mabashian** [comment]
This may be unrelated to your PR but after I save a change to a notification and get redirected to the details view, I don't see that change immediately reflected.  I have to refresh the page.  Perhaps there's some kind of race going on?

notif_details

**Comment by keithjgrant** [comment]
> This may be unrelated to your PR but after I save a change to a notification and get redirected to the details view, I don't see that change immediately reflected. I have to refresh the page. Perhaps there's some kind of race going on?

I remember this being an issue with other details pages a while back. I'll track down what we did as a fix then update this one to the same behavior

**Comment by keithjgrant** [comment]
recheck

**Comment by tiagodread** [comment]
- Trying to create  a notification without type organization and notification type (only name) The UI doesn’t show any error message

- Trying to create a notification without organization the message "This field is required." is showing twice:

> 1. Fill name
> 2. Choose notification type `Grafana`
> 3. Fill Grafana URL and Grafana API Key
> 4. Click on Save button

image

- After saving a notification on the details tab the metadata `Destination Channels` shows all destinations without any space and this happens on all the places where you can specify multiple channels or emails to receive a notification:

image

- Trying to create a notification I have to check and uncheck the `Disable SSL verification` if i want to save the form without checking this option (this also happens in all forms that use this select field):

notfcation_0

**Comment by tiagodread** [comment]
- Trying to create a notification type mattermost the field `Icon URL`must is required even the label showing not required(this also happens for rocketchat):

notfcation_1

- I don’t know why, but when you click on any label on `pagerduty` form 2x the page crashes totally and this just happens on this form (I verified all of them):

notfcation_2

Console:

`react-dom.development.js:13413 Uncaught Error: Objects are not valid as a React child (found: object with keys {_activeMessages, _catalogs, t, select, plural, selectOrdinal, _dev, _activeLanguageData, _language, _locales, _missing}). If you meant to render a collection of children, use an array instead.`

- Trying to create a notification type twilio the fields `Source phone number` and `Destination SMS number(s)` are accepting letters, is this correct ?

image

- Trying to create a email notification type, the fields `Host, Recipient list, Port and Time out` doesn’t have validation, so I can save the form like this:

image

**Review by tiagodread — CHANGES_REQUESTED** [review]
There’s some issues that we have to fix first before merging this PR

**Comment by keithjgrant** [comment]
@tiagodread I've addressed all the issues listed above, except the following items:

> Trying to create a notification without organization the message "This field is required." is showing twice

This is a known bug I plan to address in https://github.com/ansible/awx/issues/7515
I did tweak the layout a little so the first error shows in red as it is meant to and the second error sits beneath the last form field.

> After saving a notification on the details tab the metadata `Destination Channels` shows all destinations without any space and this happens on all the places where you can specify multiple channels or emails to receive a notification

I've fixed this bug in a separate forthcoming PR

> Trying to create a notification type twilio the fields Source phone number and Destination SMS number(s) are accepting letters, is this correct ?

This matches behavior of the old UI, though it certainly could be improved in the future (maybe file a low priority issue for follow-up? phone number validation is notoriously difficult to do well, made even more so by the multi-input field here)

> Trying to create a email notification type, the fields Host, Recipient list, Port and Time out doesn’t have validation

I fixed `port` and `timeout` validation, but let the other fields un-validated to match behavior of the old UI.

**Review by mabashian — APPROVED** [review]

**Comment by tiagodread** [comment]
@keithjgrant is it possible to add unique ids for these custom message inputs elements?

**Review by tiagodread — APPROVED** [review]
Retested, tests and stubs are written!
👍🏽

**Pull request merged**